Multiple Variations Support

faisal-alvi opened the pull request:
### All Submissions:

<!-- Mark completed items with an [x] -->
* [ ] Does your code follow the [WooCommerce Sniffs](https://github.com/woocommerce/woocommerce-sniffs/) variant of WordPress coding standards?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?
* [ ] Will this change require new documentation or changes to existing documentation?

---

### Changes proposed in this Pull Request:

This PR adds support for **Multiple Variations**. 

### Square and WooCommerce Attribute Types
- **Square** options:
  - **Dynamic Options**: Shared across multiple items, similar to WooCommerce’s taxonomy attributes; variation names cannot be changed.
  - **Static Options**: Unique to individual items; variation names can be changed.

- **WooCommerce** attributes:
  - **Taxonomy**: Defined globally across products.
  - **Custom**: Defined on a per-product basis.

### Key Differences Between Dynamic and Static Options
Dynamic options can be shared across products and mirror WooCommerce’s taxonomy attributes, and variation names cannot be modified, as they are automatically generated by combining the attribute values. For example: “Red, Small.” Static options are unique to each item. If multiple options exist at Square, they must be dynamic.

### Syncing Logic (Overview)
1. **WooCommerce to Square Sync**:
   - Creates **dynamic options** if there are multiple WooCommerce attributes, regardless of the WooCommerce's attributes type.
   - For a single attribute, it creates a **static option** at Square.

2. **Square to WooCommerce Sync/Import**:
   - Creates taxonomy attributes if the option slug matches an existing WooCommerce taxonomy attribute.
   - If only one option is received, it checks for an existing WooCommerce taxonomy attribute or creates a custom attribute if no match is found.

### Syncing Logic for WooCommerce and Square (Detail)

**From WooCommerce to Square:**
- If WooCommerce has more than one attribute and is set as the Source of Record (SOR), this PR will create dynamic options at Square, regardless of whether the attributes are taxonomy or custom. This is because we cannot create more than one static options for the single item (i.e product) at Square.
- If there’s only one attribute, it creates a static option at Square.
- If an attribute requires syncing, the PR creates an option at Square. New attribute values are also created at Square when synced from WooCommerce.
  
**From Square to WooCommerce:**
- If the option name received from Square matches the slug of an existing WooCommerce taxonomy attribute, this PR creates a taxonomy attribute. If the merchant updates the slug in WooCommerce, it may not match next time and a custom attribute will be created instead.
- If only one option is received, this PR checks against existing WooCommerce taxonomy attributes and attaches it if matched; otherwise, it creates a custom attribute at the product level. When creating a custom attribute, the PR reuses an existing custom attribute name, as Square does not provide the static option name. If no matching name exists, "Attribute" is assigned as the custom name.

### Square API Limitations and Solutions
Square’s API only provides the option ID used in variations, not the display name. However, to match WooCommerce taxonomy attributes or to create a custom attribute, we need the option's display name. To address this, we fetch all options via another Square API ([discussion link](https://github.com/woocommerce/woocommerce-square/issues/6#issuecomment-1881295200)) and store the option names and their IDs in a transient (`wc_square_options_data`) with a 1-day expiration. The `fetch_options_data()` function initiates this before any import or sync process.

If a transient doesn’t have updated data, a sync may fail when creating an option with a name that already exists at Square, which would require a latest options data. This PR then stops the sync and sets the `woocommerce_square_refresh_sync_cycle` option to refresh the data.

### Avoiding Attribute Name Conflicts
If WooCommerce’s taxonomy attribute is used, this PR passes the taxonomy name (e.g., “pa_color” rather than “Color”) to Square, preventing conflicts with custom attributes of the same name. The “pa_” prefix helps WooCommerce users distinguish between taxonomy and custom attributes, avoiding potential conflicts.

### New Option for Tracking
This PR introduces a new `_dynamic_options` option with `true/false` values to track whether dynamic options are used for a product. This is not in use but may be useful in future.

### Defining Variations with “Any” attributes

Noting that there's a [Defining Variations with “Any” attributes](https://woocommerce.com/document/variable-product/#variations-with-undefined-attributes) docs page for Variable Products on Woo.com that provides additional context on how this PR attempts to handle the update between WooCommerce and Square.

### Notes for Merchants
- Avoid leaving any variation dropdowns set to “Any” in products with multiple attributes, as Square does not support empty values for dynamic options. <- this one is also handled in this PR so that plugin creates "Any" option item at Square and maintains to set "Any" (means no selection) at WooCommerce and vice versa. It was decided on slack [here](https://fueled10up.slack.com/archives/C05REAQUK28/p1734108751498199?thread_ts=1733309599.907139&cid=C05REAQUK28).
- Avoid changing taxonomy attribute slugs or attribute names in WooCommerce, or renaming dynamic options in Square.
- Avoid creating custom attributes with names identical to existing taxonomy attributes, it results in issues during import/sync. Products using the same-named custom attribute are updated (replaced) incorrectly with the taxonomy attributes.
- Do not create custom attributes with names identical to existing taxonomy attributes.

### Post-Tasks

- [#251](https://github.com/woocommerce/woocommerce-square/issues/251) - Display Notices for Sync Limitations.
- [#252](https://github.com/woocommerce/woocommerce-square/issues/252) - Importing/Syncing from Square to Woocommerce, all the dynamic option values are added to the attribute, for example, for colour attribute, if there is only red and green options are selected at Square, after the import/sync, it adds all of the dynamic option values to the Woocommerce attribute regardless of what selection. This is not creating any issue to the purchases of the side, but this should be addressed.
- [#254](https://github.com/woocommerce/woocommerce-square/issues/254) - If merchant changes the name of the attribute, this will create a new option/attribute instead of using the previous one. Think of a way to handle this.

Closes #6, Closes #31 and Closes #231.

### Steps to test the changes in this Pull Request:

1. Set WooCommerce as SOR.
2. Create the following product types in WooCommerce:
   - Simple Product
   - Variable Product with:
     - 1 Custom Attribute
     - 1 Taxonomy Attribute
     - 2 Taxonomy Attributes
     - 2 Custom Attributes
     - 1 Custom + 1 Taxonomy Attribute
     - 2 Taxonomy Attributes + 1 Custom Attribute
3. Confirm these products are **auto-synced** and correctly created in Square.
4. Confirm the same for **manual sync** using the “Sync” button in settings.
5. Import all products from Square to WooCommerce and verify they retain attributes without issues (e.g., taxonomy attribute not converted to custom).
6. Make changes to the random products and confirm it syncs to the Square and Imports back to the Woo without and issues, few examples:
    1. add one more custom attribute to the “2 Taxonomy Attributes” product 
    2. remove 1 taxonomy attribute from the “2 Taxonomy Attributes + 1 Custom Attribute” product 
    3. add one more variation to any product 
    4. remove one variation from any product 
    5. change the dropdown of any variation in “1 Custom Attribute” product
    6. change the dropdown of all variations in “2 Taxonomy Attributes + 1 Custom Attribute” product
    7. add one more attribute value and create update the dropdown values of each variation
    8. etc.
7. Set Square as SOR and perform all of the above steps (sorry :)) This time the chagnes from the point 6 should be made at Square.
8. **Important**: Confirm no issues affect existing users by creating all product types in trunk and performing sync/import in the fix branch.

### Changelog entry

> Add - Support for multiple Product Variations sync.

faisal-alvi committed:
fetch options data

faisal-alvi committed:
prepare options data

faisal-alvi committed:
Stop if there are no options

faisal-alvi committed:
remove unnecessary older attributes data

faisal-alvi committed:
prepare variation data

faisal-alvi committed:
use the newer version of variation data

faisal-alvi committed:
Create transient only when step completes

faisal-alvi committed:
adding since and throw in doc

faisal-alvi committed:
vertically inlined equals

faisal-alvi committed:
do not stop if options do no exist, we may have custom variations

faisal-alvi committed:
remove extract options data logic

faisal-alvi committed:
prepare attributes for custom variations

faisal-alvi committed:
add `extract_attributes_from_square_options()`

faisal-alvi committed:
add `extract_attributes_from_square_variations()`

faisal-alvi committed:
a few spacing adjustments

faisal-alvi committed:
extract attributes as per the option type

faisal-alvi committed:
check for options just before when its needed

faisal-alvi committed:
pass `custom_meta` in data for product

faisal-alvi committed:
handle `custom_meta` and update meta

faisal-alvi committed:
update the meta name to hide it

faisal-alvi committed:
init attribute and variations

faisal-alvi committed:
consider dynamic options if more than one attributes exits; also delete the dynamic option meta

faisal-alvi committed:
Set the product as a dynamic options product

faisal-alvi committed:
Loop through the attributes to create options and values at Square.

faisal-alvi committed:
Set the item_option_id for each option to the product

faisal-alvi committed:
Check if Square already has the option created with the same name

faisal-alvi committed:
If name does not exist, create a new option in Square

faisal-alvi committed:
Loop through the attribute values to create option values

faisal-alvi committed:
Set the option values

faisal-alvi committed:
Push option object to Square to create a new one. Used timestamp as idempotency_key

faisal-alvi committed:
Handle error

faisal-alvi committed:
Log the error and throw it

faisal-alvi committed:
Add @todo list

faisal-alvi committed:
Set the name of the variation if it's a single variation

faisal-alvi committed:
Handle dynamic options while syncing to Square

faisal-alvi committed:
Set the `item_option_values` for the variation

faisal-alvi committed:
Prepare data if data found in transient

faisal-alvi committed:
Create step to fetch_options_data in Manual Sync

faisal-alvi committed:
Request refresh when we do not have Square's Dynamic options data ready

faisal-alvi committed:
Update the fetch_options_data function to be dynamically used in more files

faisal-alvi committed:
Remove unused namespace

faisal-alvi committed:
Add a function `retrieve_options_data` to be used in more files

faisal-alvi committed:
Use `ListCatalogResponse`

faisal-alvi committed:
Handle the new cursor properly

faisal-alvi committed:
fix variable name

faisal-alvi committed:
fix spacing

faisal-alvi committed:
Fix: Taxonomy Attribute converting to the Custom Attribute

faisal-alvi committed:
Fix: attribute name values changed at square issue!

faisal-alvi committed:
Fix: 'Attribute' set as attribute name issue

faisal-alvi committed:
remove use of catalog_options

faisal-alvi committed:
imrpove condition; skip checking catalog options

faisal-alvi committed:
options_ID to options_id

faisal-alvi committed:
get attribute values as per the type

faisal-alvi committed:
check if attribute and its all values exist, if not create!

faisal-alvi committed:
reset options when product has only one attribute

faisal-alvi committed:
also for simple products, reset options from catalog

faisal-alvi committed:
use correct name for attribute

faisal-alvi committed:
place init vars on correct location

faisal-alvi committed:
use attribute_id to match with transient name

faisal-alvi committed:
check for attribute value's Square ID

faisal-alvi committed:
remove loop break for both IDs found

faisal-alvi committed:
create options and values when IDs not found in transient

faisal-alvi committed:
variation name as combination of all attribute values

faisal-alvi committed:
pass slug to find the taxonomy

faisal-alvi committed:
save option value ids in transient

faisal-alvi committed:
again pass slug to catch the taxonomy

faisal-alvi committed:
replace 'pa_' from attribute name to use as a slug

faisal-alvi committed:
API function to create options and values at Square

faisal-alvi committed:
Merge branch 'trunk' into multiple-variations-support

faisal-alvi committed:
phpcs: spacing issues fixed

faisal-alvi committed:
phpcs fix about the short ternaries
includes/Sync/Product_Import.php
-----------------------------------------------------------------------------------------------
FOUND 1 ERROR AND 0 WARNINGS AFFECTING 1 LINE
-----------------------------------------------------------------------------------------------
 92 | ERROR | Using short ternaries is not allowed as they are rarely used correctly (Universal.Operators.DisallowShortTernary.Found)
-----------------------------------------------------------------------------------------------

faisal-alvi committed:
phpcs fix for short ternary in another file
FILE: includes/Sync/Manual_Synchronization.php
-----------------------------------------------------------------------------------------------
FOUND 1 ERROR AND 0 WARNINGS AFFECTING 1 LINE
-----------------------------------------------------------------------------------------------
 1684 | ERROR | Using short ternaries is not allowed as they are rarely used correctly (Universal.Operators.DisallowShortTernary.Found)
-----------------------------------------------------------------------------------------------

faisal-alvi committed:
phpcs fixes for variable naming, params, etc
FILE: includes/Handlers/Product/Woo_SOR.php
-----------------------------------------------------------------------------------------------
FOUND 14 ERRORS AND 0 WARNINGS AFFECTING 14 LINES
-----------------------------------------------------------------------------------------------
  87 | ERROR | Using short ternaries is not allowed as they are rarely used correctly (Universal.Operators.DisallowShortTernary.Found)
 113 | ERROR | The parameter "array( 'hide_empty' => false )" at position #2 of get_terms() has been deprecated since WordPress version 4.5.0. Instead do not pass the parameter. (WordPress.WP.DeprecatedParameters.Get_termsParam2Found)
 142 | ERROR | Variable "$CatalogItemOptionForItem" is not in valid snake_case format, try "$catalog_item_option_for_item" (WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase)
 143 | ERROR | Variable "$CatalogItemOptionForItem" is not in valid snake_case format, try "$catalog_item_option_for_item" (WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase)
 144 | ERROR | Variable "$CatalogItemOptionForItem" is not in valid snake_case format, try "$catalog_item_option_for_item" (WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase)
 348 | ERROR | Variable "$CatalogItemOptionValueForItemVariation" is not in valid snake_case format, try "$catalog_item_option_value_for_item_variation" (WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase)
 349 | ERROR | Variable "$CatalogItemOptionValueForItemVariation" is not in valid snake_case format, try "$catalog_item_option_value_for_item_variation" (WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase)
 350 | ERROR | Variable "$CatalogItemOptionValueForItemVariation" is not in valid snake_case format, try "$catalog_item_option_value_for_item_variation" (WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase)
 352 | ERROR | Variable "$CatalogItemOptionValueForItemVariation" is not in valid snake_case format, try "$catalog_item_option_value_for_item_variation" (WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase)
 357 | ERROR | The parameter "array( 'hide_empty' => false )" at position #2 of get_terms() has been deprecated since WordPress version 4.5.0. Instead do not pass the parameter. (WordPress.WP.DeprecatedParameters.Get_termsParam2Found)
 378 | ERROR | Variable "$CatalogItemOptionValueForItemVariation" is not in valid snake_case format, try "$catalog_item_option_value_for_item_variation" (WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase)
 379 | ERROR | Variable "$CatalogItemOptionValueForItemVariation" is not in valid snake_case format, try "$catalog_item_option_value_for_item_variation" (WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase)
 380 | ERROR | Variable "$CatalogItemOptionValueForItemVariation" is not in valid snake_case format, try "$catalog_item_option_value_for_item_variation" (WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase)
 382 | ERROR | Variable "$CatalogItemOptionValueForItemVariation" is not in valid snake_case format, try "$catalog_item_option_value_for_item_variation" (WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase)
-----------------------------------------------------------------------------------------------

faisal-alvi committed:
add Todo task to improve approach

faisal-alvi committed:
update doc comments

iamdharmesh committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-square into multiple-variations-support

iamdharmesh committed:
Added initial tests.

iamdharmesh committed:
Disable new product editor in sync tests.

faisal-alvi committed:
Multiple Vars Support in Manual Sync

faisal-alvi committed:
change the func scope to avail it for sync process

faisal-alvi committed:
Multiple Vars Support in Auto Sync / Interval Polling

faisal-alvi committed:
apply filters for name and desc on update

faisal-alvi committed:
remove unnecessary overwrites by old methods

faisal-alvi committed:
if no variation was found, check if the parent product exists

faisal-alvi committed:
fix issue of 0 option values expected, found 3

faisal-alvi committed:
fix attributes object vs variation items

faisal-alvi committed:
fixed: [BAD_REQUEST] Expected ItemVariation to have 0 Item Option Values, got 2

dkotter commented on includes/Handlers/Product/Woo_SOR.php:
```diff
@@ -81,12 +81,75 @@ public static function update_catalog_item( CatalogObject $catalog_object, \WC_P
+				// Loop through the attributes to create options and values at Square.
+				foreach ( $attributes as $attribute_id => $attribute ) {
+
+					$attribute_name = $attribute->get_name(); //wow it gives pa_size which saved me!
```
Assuming we don't need this comment:

```suggestion
					$attribute_name = $attribute->get_name();
```

dkotter commented on includes/Handlers/Product/Woo_SOR.php:
```diff
@@ -225,7 +290,102 @@ public static function update_catalog_variation( CatalogObject $catalog_object,
+						$taxonomy_exists  = true;
+					} else {
+						// For custom attributes, simply use the cleaned-up attribute ID
+						$attribute_name   = ucwords( str_replace( '-', ' ', $attribute_id ) ); // what if merchant changes the name?
```
Maybe change this to a `TODO` comment to make it easier to find in the future? Or remove all together if we have this tracked elsewhere

faisal-alvi replied:
Removed it from here as this would be handled in https://github.com/woocommerce/woocommerce-square/issues/254, thanks!

dkotter commented on includes/Sync/Product_Import.php:
```diff
@@ -874,6 +1066,13 @@ protected function save_product_meta( $product_id, $data ) {
+		// Update custom meta.
+		if ( $data['custom_meta'] ) {
+			foreach ( $data['custom_meta'] as $meta_key => $meta_value ) {
+				update_post_meta( $product_id, $meta_key, $meta_value );
```
Do we need any sanitization on these values before updating the meta? Or is that done upstream?

faisal-alvi replied:
This seems to be sanitized in core via `sanitize_meta()`.

faisal-alvi committed:
removing comment

Co-authored-by: Darin Kotter <darin.kotter@gmail.com>

iamdharmesh committed:
Merge branch 'trunk' into multiple-variations-support

iamdharmesh committed:
Add Additional multiple variations tests.

iamdharmesh committed:
Add tests for update product.

faisal-alvi committed:
Removed it from here as this would be handled in issue 254

faisal-alvi committed:
Merge branch 'trunk' into multiple-variations-support

iamdharmesh committed:
Merge push tests into single tests to save test run time.

iamdharmesh committed:
Added initial tests for Square SOR sync.

iamdharmesh committed:
Merge branch 'multiple-variations-support' of github.com:woocommerce/woocommerce-square into multiple-variations-support

iamdharmesh committed:
Tag the test correctly.

iamdharmesh commented:
Hi @faisal-alvi, 

I have added initial E2E tests for the Woo SOR with products created using the following combinations:
>- Variable Product with:
>     - 1 Custom Attribute
>     - 1 Taxonomy Attribute
>     - 2 Taxonomy Attributes
>     - 2 Custom Attributes
>     - 1 Custom + 1 Taxonomy Attribute
>     - 2 Taxonomy Attributes + 1 Custom Attribute

Additionally, I have started writing E2E tests for the Square SOR. However, while working on it, I noticed an issue: if an attribute name begins with `pa_` and a global attribute with the same name does not exist, the attribute value does not get selected in the created variations. Please refer to the screenshots below for clarification:

| Square Side | WC Side |
| - | - |
| ![image](https://github.com/user-attachments/assets/460b376b-a573-4fff-881d-d71aa50dd555) | ![image](https://github.com/user-attachments/assets/8c550793-8fd5-46ad-b5bd-0a5d1e99509c) |

**Steps to Reproduce:**
1. Delete all product attributes.
1. Create item options in Square with the names `pa_color` and `pa_size`, and add option values to these attributes.
1. Create a product on the Square using the created item options.
1. Import the product and check the variations in the imported product.

_If you don't want to run these steps you can simply run the test `[Square SOR] Import multiple variations products from Square` from `d8.square-sor-multiple-variations.spec.js` file and can see the product variations of the product._

Please let me know if you need any further information on this.

Thank you.

iamdharmesh committed:
Update time limit for looking for catalog data.

iamdharmesh committed:
Fix flaky tests of Square SOR

faisal-alvi committed:
create new taxonomy attribute if does not found on import
(for `pa_` prefixed names only)

faisal-alvi commented:
Hi @iamdharmesh, Thank you for the detailed comment and the E2E tests—much appreciated! Regarding the issue you mentioned, the presence of the `pa_` prefix in Square typically means the option was synced from WooCommerce. This scenario can occur if the merchant deletes the corresponding taxonomy in WooCommerce after syncing.  

I've addressed this. Now, in such cases, a plugin will automatically create a new taxonomy attribute in WooCommerce if one with the `pa_` prefix doesn't exist and assign it accordingly. Please re-run the tests on your end and let me know if they work. Also, feel free to share if you notice anything else while working on the E2E tests. Once E2E tests are finished, please move the ticket further on the board. Thanks!

iamdharmesh committed:
Update Square SOR test.

faisal-alvi committed:
If the attribute value is 'Any', add it to the attribute option values

faisal-alvi committed:
Set empty if attribute is to 'Any' to prevent it from being saved

faisal-alvi committed:
fix phps errors,
FILE: includes/Sync/Product_Import.php
-----------------------------------------------------------------------------------------------
FOUND 4 ERRORS AND 1 WARNING AFFECTING 5 LINES
-----------------------------------------------------------------------------------------------
  982 | WARNING | Usage of ELSE IF is discouraged; use ELSEIF instead (PSR2.ControlStructures.ElseIfDeclaration.NotAllowed)
  989 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
  999 | ERROR | All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found '$attribute_id'. (WordPress.Security.EscapeOutput.ExceptionNotEscaped)
 1012 | ERROR | Spaces must be used for mid-line alignment; tabs are not allowed (Universal.WhiteSpace.DisallowInlineTabs.NonIndentTabsUsed)
 1012 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
-----------------------------------------------------------------------------------------------

faisal-alvi committed:
phpcs spacing issue fixed,
FILE: includes/Sync/Product_Import.php
-----------------------------------------------------------------------------------------------
FOUND 1 ERROR AND 0 WARNINGS AFFECTING 1 LINE
-----------------------------------------------------------------------------------------------
 989 | ERROR | Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
-----------------------------------------------------------------------------------------------

iamdharmesh committed:
TEMP: try identify failure root cause.

iamdharmesh committed:
Try update idempotency key to something better, instead of `time()` which tend to fail.

iamdharmesh committed:
Revert "TEMP: try identify failure root cause."

This reverts commit ed806119798ef1e75604f0fd9f6fc72c09de1657.

iamdharmesh commented on includes/Handlers/Product/Woo_SOR.php:
```diff
@@ -81,12 +81,75 @@ public static function update_catalog_item( CatalogObject $catalog_object, \WC_P
+				count( $attributes ) > 1
+			) {
+				$options_ids            = array();
+				$options_data_transient = get_transient( 'wc_square_options_data' );
```
I think we should have a fallback mechanism to generate the requested data in case it is not available in the transient storage. If generating the data is too heavy, we may need to consider saving the data in a more reliable location. This is because there is no guarantee that transient data will persist until its expiration time, it might disappear just seconds after being set. More information [here](https://developer.wordpress.org/apis/transients/#overview:~:text=Everyone%20seems%20to%20misunderstand%20how%20transient%20expiration%20works%2C%20so%20the%20long%20and%20short%20of%20it%20is%3A%20transient%20expiration%20times%20are%20a%20maximum%20time.%20There%20is%20no%20minimum%20age.%20Transients%20might%20disappear%20one%20second%20after%20you%20set%20them%2C%20or%2024%20hours%2C%20but%20they%20will%20never%20be%20around%20after%20the%20expiration%20time.)

iamdharmesh commented:
@faisal-alvi The E2E test was failing due to the idempotency key being reused for the upsert item option. I have updated the logic of key generation [here](https://github.com/woocommerce/woocommerce-square/pull/235/commits/5c1644a76f853942e325f2773fcd9b277edd29fc) and its working fine now. Please feel free to look and update if needed. Thanks.

faisal-alvi committed:
Merge branch 'trunk' into multiple-variations-support

faisal-alvi committed:
use options instead of transients

faisal-alvi committed:
Merge branch 'multiple-variations-support' of github.com:woocommerce/woocommerce-square into multiple-variations-support

faisal-alvi committed:
use transient in a more safer way

faisal-alvi commented:
@dkotter over to you for code review.

vikrampm1 commented:
@qasumitbagthariya bumping this up in your queue to try and wrap it up asap.

faisal-alvi committed:
Merge branch 'trunk' into multiple-variations-support

qasumitbagthariya commented:
### QA Update  ✅ 


-------------

I have verified this PR in the `multiple-variations-support` branch which has been fixed and is functioning as intended.

I tested the following on this branch:
Verified all the cases related to variations and the steps provided in the PR description

**Square SOR**

https://github.com/user-attachments/assets/c2013784-28c3-40f6-a2a5-253ea8de349d

**Woocommerce SOR**

https://github.com/user-attachments/assets/cfb1c89e-1e17-4a7c-9efe-82212d38fae1



**Testing Environment** 

<Details>

- WordPress: 6.7.1
- Theme: Twenty Twenty-Four 1.3
- WooCommerce - 9.5.1
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS 15.2 
- Branch: multiple-variations-support

</Details>


Steps to Test- As mentioned in the PR description.
Test Results - It is working as expected.
Functional Demo / Screencast - 
Special Notes -  Ready for code review (Woo)
Testing Document status:
Cases related to this Issue/PR are added to the Critical Flow Wiki pages: 

- [x] Yes
- [ ] Not Required/Applicable for this PR
------------------------------------------------

wjrosa commented on includes/API.php:
```diff
@@ -597,6 +598,159 @@ public function retrieve_inventory_physical_count( $physical_count_id ) {
+		$objects = $response->get_data()->getObjects() ? $response->get_data()->getObjects() : array();
+
+		foreach ( $objects as $object ) {
+			$options_data[ $object->getId() ]['name'] = $object->getItemOptionData()->getDisplayName();
```
Can any of the getters return `null` values? If so, it would be nice to add some checks to avoid errors.

faisal-alvi replied:
At this point in the code, there is a `getDisplayName`, which appears to be present for every item.

wjrosa commented on includes/Handlers/Product/Woo_SOR.php:
```diff
@@ -81,12 +81,76 @@ public static function update_catalog_item( CatalogObject $catalog_object, \WC_P
+		$attributes = $product->get_attributes();
+
+		$product_variation_ids = $product->get_children();
+		$catalog_variations    = $item_data->getVariations() ? $item_data->getVariations() : array();
```
Since PHP 7.0 we can change this to:
```suggestion
		$catalog_variations    = $item_data->getVariations() ?? array();
```

wjrosa commented on includes/Handlers/Product/Woo_SOR.php:
```diff
@@ -81,12 +81,76 @@ public static function update_catalog_item( CatalogObject $catalog_object, \WC_P
+			) {
+				$options_ids  = array();
+				$result       = wc_square()->get_api()->retrieve_options_data();
+				$options_data = isset( $result[1] ) ? $result[1] : array();
```
Same as above:
```suggestion
				$options_data = $result[1] ?? array();
```

wjrosa commented on includes/Handlers/Product/Woo_SOR.php:
```diff
@@ -225,7 +291,111 @@ public static function update_catalog_variation( CatalogObject $catalog_object,
+			$variation_items       = $product->get_attributes();
+			$variation_item_values = array();
+
+			if ( 1 === count( $attributes ) ) {
```
I see that these changes contain multiple "actions" that could be new methods to make reading easier. Basically, every time we have a comment here, the following code could be a new method named after the explanation. But this is optional

wjrosa commented on includes/Sync/Manual_Synchronization.php:
```diff
@@ -1698,6 +1743,25 @@ protected function assign_next_steps() {
+		$new_cursor = isset( $result[2] ) ? $result[2] : null;
+
+		if ( ! empty( $new_cursor ) ) {
+			$this->set_attr( 'fetch_options_data_cursor', $new_cursor );
```
I believe this line can be moved outside the condition, as `$new_cursor` will be null per the default value above anyway.

wjrosa commented on includes/Sync/Product_Import.php:
```diff
@@ -1056,6 +1297,8 @@ protected function save_variations( $product_id, $data ) {
 
 				foreach ( $variation['attributes'] as $attribute_key => $attribute ) {
+					// Set empty if attribute is to 'Any' to prevent it from being saved.
+					$attribute['option'] = isset( $attribute['option'] ) && 'Any' !== $attribute['option'] ? $attribute['option'] : '';
```
Maybe we should move `Any` to a new constant.

wjrosa reviewed (approved):
Looks good! Some multiple optional suggestions 👍

vikrampm1 commented:
Thanks @wjrosa. @faisal-alvi see if you can make those optional suggested changes above and then we can move this to the next step.

jeffpaul commented:
Noting that there's a [Defining Variations with “Any” attributes](https://woocommerce.com/document/variable-product/#variations-with-undefined-attributes) docs page for Variable Products on Woo.com that provides additional context on how this PR attempts to handle the update between WooCommerce and Square.

faisal-alvi committed:
Merge branch 'trunk' into multiple-variations-support

faisal-alvi committed:
suggestions implemented

wjrosa reviewed (approved):
Thanks for addressing the suggestions! Looks good to me 👍

jeffpaul commented:
@vikrampm1 @faisal-alvi documenting some items here from our conversation with Woo & Square on 29 JAN:
* [ ] Square to craft a video tutorial for Multiple Variations to be used in merchant education / marketing
* [ ] Chris/Kevin @ Woo will look to help with merchant outreach for feedback
* [ ] Square & Woo account management & customer support to check on merchant feedback
* [x] 10up to ping Woo HEs for pre-existing merchant request feedback (done in https://github.com/woocommerce/woocommerce-square/issues/6#issuecomment-2622202560)

dkotter committed:
Merge branch 'trunk' into multiple-variations-support

faisal-alvi committed:
Merge branch 'trunk' into multiple-variations-support

jeffpaul commented:
@faisal-alvi (and others watching this PR) note that we're awaiting confirmation to release these changes as-is and will advance this for Regression Testing once that's confirmed.

qasumitbagthariya commented:
Regression / Smoke Test Report  ✅ 
-------

Tested with Archive File created via "php woorelease.phar build repo_URL" (Composer version 2.8.4, npm version 10.8.2, node version 20.18.1)

Status- Working expected with Plugin Archive/Zip file same as fix specific branch.

**Testing Environment** 

<Details>

- WordPress: 6.7.1
- Theme: Storefront 4.6.0
- Theme: Twenty Twenty-Four 1.3
- WooCommerce - 9.7.1
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS 15.2 
- Branch: smoke-testing

</Details>


Next Step- Ready to Merge 🚀
